### PR TITLE
Issue 170: Alert Distribution

### PR DESCRIPTION
**Pull request by cAbhi15**
Linked to issue(s): #170 

## What changes were proposed in this pull request?
- adds the distribution pipeline:
  - read the science db (Hbase)
  - serialize into avro and publish to Kafka
  - a Kafka consumer to deserialize the avro message and do spark DF operations
- script for managing the Kafka server

## To-do

- [x] integration test for the distribution pipeline

- [x] schema for distribution

## How was this patch tested?
- manually and CI

### CODE REVIEW

File: bin/distribute.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        .load()
+
+    # Get the DataFrame for publishing to Kafka (avro serialized)
+    df_kafka = get_kafka_df(df, args.distribution_schema)
+
+    # Publish to a test topic (Ensure that the topic exists on the Kafka Server)
+    topic = "distribution_test"
+
+    df_kafka\
+        .write\
```
Comment (JulienPeloton) [inline bin/distribute.py]: The distribution services stops after sending all alerts as you did not make it a streaming service for the moment. Actually, I do not know what will be the real use in the future (calling it manually when needed, or having a streaming service publishing automatically alerts), but could you add an option and extend the script to run both cases?

File: bin/fink_kafka [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+message_help="""
+Manage Fink's Kafka Server for Alert Distribution \n
+USAGE:\n\n
```
Comment (JulienPeloton) [inline bin/fink_kafka]: Minor - could you use the same indentation as is used for the `fink` & `fink_test` binaries?

File: bin/raw2science.py [hunk 3]
```diff
@@ -87,8 +87,8 @@ def main():
     catalog = construct_hbase_catalog_from_flatten_schema(
         df_hbase.schema, args.science_db_name, "objectId")
 
-    # Save the catalog on disk for later usage
-    with open('catalog.json', 'w') as json_file:
+    science_db_catalog = args.science_db_catalog
+    with open(science_db_catalog, 'w') as json_file:
```
Comment (JulienPeloton) [inline bin/raw2science.py]: Good catch ;-)

File: conf/fink.conf [hunk 4]
```diff
@@ -98,3 +99,9 @@ TIME_INTERVAL=0.1
 
 # Total number of alerts to send
 POOLSIZE=100
+
+######################################
+# Distribution
+# Configurations for Alert Re-Distribution
+# The path where to store the avro distribution schema
+DISTRIBUTION_SCHEMA=${FINK_HOME}/schemas/distribution_schema
```
Comment (JulienPeloton) [inline conf/fink.conf]: Could you add an extension to the file? I find it easier to know what it is.

File: conf/fink.conf.cluster [hunk 5]
```diff
@@ -98,3 +98,9 @@ TIME_INTERVAL=0.1
 
 # Total number of alerts to send
 POOLSIZE=100
+
+######################################
+# Distribution
+# Configurations for Alert Re-Distribution
+# The path where to store the avro distribution schema
+DISTRIBUTION_SCHEMA=${FINK_HOME}/schemas/distribution_schema
```
Comment (JulienPeloton) [inline conf/fink.conf.cluster]: Could you add an extension to the file? I find it easier to know what it is.

File: conf/fink.conf.cluster [hunk 6]
```diff
@@ -98,3 +98,9 @@ TIME_INTERVAL=0.1
```
Comment (JulienPeloton) [inline conf/fink.conf.cluster]: Missing `SCIENCE_DB_CATALOG=${FINK_HOME}/catalog.json` in this conf file.

File: conf/fink.conf.travis [hunk 7]
```diff
@@ -98,3 +99,9 @@ TIME_INTERVAL=0.1
 
 # Total number of alerts to send
 POOLSIZE=10
+
+######################################
+# Distribution
+# Configurations for Alert Re-Distribution
+# The path where to store the avro distribution schema
+DISTRIBUTION_SCHEMA=${FINK_HOME}/schemas/distribution_schema
```
Comment (JulienPeloton) [inline conf/fink.conf.travis]: Could you add an extension to the file? I find it easier to know what it is.

File: docs/user_guide/streaming-out.md [hunk 8]
```diff
@@ -1,3 +1,82 @@
...
+
+1. Kafka Cluster
+2. Spark Process (acts as the Producer)
+
+## Kafka Cluster
+The Kafka Cluster consists of one or more Kafka broker(s) and a Zookeeper server that monitors and manages the Kafka brokers.
+It is important to note that the Kafka Cluster for Redistribution of alerts is different from the one used in [simulator](simulator.md).
+<br>
+You will need [Apache Kafka](https://kafka.apache.org/) 2.2+ installed. Define `KAFKA_HOME` as per your installation in your ~/.bash_profile.
+
```
Comment (JulienPeloton) [inline docs/user_guide/streaming-out.md]: Maybe linking here to the installation script `conf/install_kafka.sh`.

File: bin/distribute.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+import argparse
+import json
+
+from fink_broker.parser import getargs
+from fink_broker.sparkUtils import init_sparksession
+from fink_broker.distributionUtils import get_kafka_df
+
+def main():
+    parser = argparse.ArgumentParser(description = __doc__)
```
Comment (JulienPeloton) [inline bin/distribute.py]: PEP8: no spaces around keywords

File: bin/distribute.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    args = getargs(parser)
+
+    # Get or create a Spark Session
+    spark = init_sparksession(
+        name = "distribution", shuffle_partitions = 2, log_level = "ERROR")
```
Comment (JulienPeloton) [inline bin/distribute.py]: PEP8: no spaces around keywords

File: bin/distribute.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    # Publish to a test topic (Ensure that the topic exists on the Kafka Server)
+    topic = "distribution_test"
+
+    df_kafka\
+        .write\
+        .format("kafka")\
+        .option("kafka.bootstrap.servers", "localhost:9093")\
+        .option("topic", topic)\
+        .save()
+
```
Comment (JulienPeloton) [inline bin/distribute.py]: PEP8: 2 blank lines before the main

File: bin/distribution_test.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+import argparse
+
+from fink_broker.parser import getargs
+from fink_broker.sparkUtils import init_sparksession
+from fink_broker.distributionUtils import decode_kafka_df
+from pyspark.sql.functions import col
+
+def main():
+    parser = argparse.ArgumentParser(description = __doc__)
```
Comment (JulienPeloton) [inline bin/distribution_test.py]: PEP8: no spaces around keywords

File: bin/distribution_test.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    args = getargs(parser)
+
+    # Get or create a Spark Session
+    spark = init_sparksession(
+        name = "consume_distribution", shuffle_partitions = 2, log_level = "ERROR")
```
Comment (JulienPeloton) [inline bin/distribution_test.py]: PEP8: no spaces around keywords

File: bin/distribution_test.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        .read \
+        .format("kafka") \
+        .option("kafka.bootstrap.servers", "localhost:9093") \
+        .option("value.deserializer", "ByteArrayDeserializer") \
+        .option("subscribe", topic) \
+        .load()
+
+    # Decode df_kafka into a Spark DataFrame with StructType column
+    df = decode_kafka_df(df_kafka, args.distribution_schema)
+
```
Comment (JulienPeloton) [inline bin/distribution_test.py]: PEP8: Too many blank lines (1 only)

File: bin/distribution_test.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    print("\nprinting the schema of the decoded df\n")
+    df.printSchema()
+
+    print("\nPrint first three rows\n")
+    df.show(3)
+    df.select(col("struct.*")).show(1)
+
```
Comment (JulienPeloton) [inline bin/distribution_test.py]: PEP8: 2 blank lines before the main

File: conf/kafka.server.properties [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
+############################# Log Retention Policy #############################
+
+# The following configurations control the disposal of log segments. The policy can
+# be set to delete segments after a period of time, or after a given size has accumulated.
+# A segment will be deleted whenever *either* of these criteria are met. Deletion always happens
+# from the end of the log.
+
+# The minimum age of a log file to be eligible for deletion due to age
+log.retention.hours=168
```
Comment (JulienPeloton) [inline conf/kafka.server.properties]: Maybe retention period of 168h is quite big by default, as space fills up pretty quickly with tests ;-) 

I usually find 24h more suitable (then users can update for their need if they know they will have plenty of free space).

File: docs/user_guide/streaming-out.md [hunk 17]
```diff
@@ -1,3 +1,82 @@
...
+# in ~/.bash_profile
+# as per your Kafka installation directory
+export KAFKA_HOME=/usr/local/kafka
+```
+<br>
+We provide a script `fink_kafka` to efficiently manage the Kafka Cluster for Redistribution. The help message shows the available
+services:
+
+```bash
+Manage Finks Kafka Server for Alert Distribution
```
Comment (JulienPeloton) [inline docs/user_guide/streaming-out.md]: Indentation to be updated with comment above.

File: python/fink_broker/distributionUtils.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+import glob
+import shutil
+
+from fink_broker.avroUtils import readschemafromavrofile
+from fink_broker.sparkUtils import get_spark_context, to_avro, from_avro
+from pyspark.sql import DataFrame
+from pyspark.sql.functions import struct, col
+from fink_broker.tester import spark_unit_tests
+
+def get_kafka_df(df : DataFrame, schema_path : str) -> DataFrame:
```
Comment (JulienPeloton) [inline python/fink_broker/distributionUtils.py]: PEP8: no whitespace before `:`

File: python/fink_broker/distributionUtils.py [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+    # Convert into avro
+    df_kafka = df_struct.select(to_avro("struct").alias("value"))
+
+    # Store the avro schema if it doesn't already exist
+    # NOTE:
+    #   Ensure to delete the avro schema at schema_path whenever the structure
+    #   of the df read from science db or the contents to be distribution needs
+    #   to be changed.
+
+    if not os.path.isfile(schema_path):
```
Comment (JulienPeloton) [inline python/fink_broker/distributionUtils.py]: Maybe moving this part into another function for clarity.

File: python/fink_broker/distributionUtils.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+        # Write the schema to a file for decoding Kafka messages
+        with open(schema_path, 'w') as f:
+            json.dump(avro_schema, f)
+
+        # Remove .avro files and directory
+        shutil.rmtree(path_for_avro)
+
+    return df_kafka
+
+def decode_kafka_df(df_kafka: DataFrame, schema_path: str) -> DataFrame:
```
Comment (JulienPeloton) [inline python/fink_broker/distributionUtils.py]: PEP8: no whitespace before `:`

File: python/fink_broker/sparkUtils.py [hunk 21]
```diff
@@ -61,6 +62,45 @@ def from_avro(dfcol: Column, jsonformatschema: str) -> Column:
...
+    >>> avro_example_schema = '''
+    ... {
+    ...     "type" : "record",
+    ...     "name" : "struct",
+    ...     "fields" : [
+    ...             {"name" : "col1", "type" : "long"},
+    ...             {"name" : "col2", "type" : "string"}
+    ...     ]
+    ... }'''
+    >>> df = spark.range(5).select(struct("id", col("id").cast("string").alias("id2")).alias("struct"))
```
Comment (JulienPeloton) [inline python/fink_broker/sparkUtils.py]: Could you break the line to make it readable on small screen?

**Comment by cAbhi15** [comment]
Thanks @JulienPeloton for the detailed review. I will update the PR with the requested changes soon.

File: conf/fink.conf (hunk 4 above)
Comment (cAbhi15) [reply conf/fink.conf]: added

File: conf/fink.conf.cluster (hunk 5 above)
Comment (cAbhi15) [reply conf/fink.conf.cluster]: done

File: conf/fink.conf.cluster (hunk 6 above)
Comment (cAbhi15) [reply conf/fink.conf.cluster]: added

File: conf/fink.conf.travis (hunk 7 above)
Comment (cAbhi15) [reply conf/fink.conf.travis]: done

File: docs/user_guide/streaming-out.md (hunk 8 above)
Comment (cAbhi15) [reply docs/user_guide/streaming-out.md]: done

File: bin/distribute.py (hunk 9 above)
Comment (cAbhi15) [reply bin/distribute.py]: fixed

File: bin/distribute.py (hunk 10 above)
Comment (cAbhi15) [reply bin/distribute.py]: fixed

File: bin/distribute.py (hunk 11 above)
Comment (cAbhi15) [reply bin/distribute.py]: fixed

File: bin/distribution_test.py (hunk 12 above)
Comment (cAbhi15) [reply bin/distribution_test.py]: fixed

File: bin/distribution_test.py (hunk 13 above)
Comment (cAbhi15) [reply bin/distribution_test.py]: fixed

File: bin/distribution_test.py (hunk 14 above)
Comment (cAbhi15) [reply bin/distribution_test.py]: fixed

File: bin/distribution_test.py (hunk 15 above)
Comment (cAbhi15) [reply bin/distribution_test.py]: fixed

File: conf/kafka.server.properties (hunk 16 above)
Comment (cAbhi15) [reply conf/kafka.server.properties]: changed to 24h

File: docs/user_guide/streaming-out.md (hunk 17 above)
Comment (cAbhi15) [reply docs/user_guide/streaming-out.md]: done

File: python/fink_broker/distributionUtils.py (hunk 18 above)
Comment (cAbhi15) [reply python/fink_broker/distributionUtils.py]: fixed

File: python/fink_broker/distributionUtils.py (hunk 19 above)
Comment (cAbhi15) [reply python/fink_broker/distributionUtils.py]: moved to another function

File: python/fink_broker/distributionUtils.py (hunk 20 above)
Comment (cAbhi15) [reply python/fink_broker/distributionUtils.py]: fixed

File: python/fink_broker/sparkUtils.py (hunk 21 above)
Comment (cAbhi15) [reply python/fink_broker/sparkUtils.py]: done

File: bin/distribute.py (hunk 1 above)
Comment (cAbhi15) [reply bin/distribute.py]: Actually, I am yet to figure out how I can create a streaming service that reads the HBase as it gets updated with new data. Can I update it in another PR?

File: bin/fink_kafka (hunk 2 above)
Comment (cAbhi15) [reply bin/fink_kafka]: I have updated the indentation. It looks something like this:
```bash
Manage Fink's Kafka Server for Alert Distribution 

 Usage:
 	fink_kafka [option] 

 Option:
 	start
 		Starts Zookeeper Server and a Kafka Cluster 

 	stop 
 		Stops the running Kafka Cluster and Zookeeper Server 

 	-h, --help 
 		To view this help message 

 	--create-topic <TOPIC> 
 		Creates a topic named <TOPIC> if it does not already exist
```
Is it alright? or did you want me to change the help message to make it similar to the fink and fink_test binaries?
For e.g. like:
```bash
Manage Fink's Kafka Server for Alert Distribution 

 Usage:
 	to start: fink_kafka start
        to stop: fink_kafka stop
        to create a new topic: fink_kafka --create-topic <TOPIC>

To get this help:
        fink_kafka
        fink_kafka -h
        fink_kafka --help
```

File: bin/distribute.py (hunk 1 above)
Comment (JulienPeloton) [reply bin/distribute.py]: Good idea to move this to another PR actually!

File: bin/fink_kafka (hunk 2 above)
Comment (JulienPeloton) [reply bin/fink_kafka]: It is alright :-)

File: python/fink_broker/distributionUtils.py [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+    at schema_path which is passed as an argument.
+
+    Parameters
+    ----------
+    df: DataFrame
+        A Spark DataFrame created after reading the science database (HBase)
+    schema_path: str
+        Path where to store the avro schema required for decoding the
+        Kafka messages.
+    ----------
```
Comment (JulienPeloton) [inline python/fink_broker/distributionUtils.py]: Missing `Returns` section in the docstring.

File: python/fink_broker/distributionUtils.py [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+    avro(binary). This routine creates a Spark DataFrame with a decoded StructType
+    column using the avro schema at schema_path.
+
+    Parameters
+    ----------
+    df_kafka: DataFrame
+        A Spark DataFrame created after reading the Kafka Source
+    schema_path: str
+        Path where the avro schema to decode the Kafka message is stored
+
```
Comment (JulienPeloton) [inline python/fink_broker/distributionUtils.py]: Missing `Returns` section in the docstring.

File: python/fink_broker/distributionUtils.py (hunk 22 above)
Comment (cAbhi15) [reply python/fink_broker/distributionUtils.py]: fixed in 4a89224

File: python/fink_broker/distributionUtils.py (hunk 23 above)
Comment (cAbhi15) [reply python/fink_broker/distributionUtils.py]: fixed in 4a89224

**Comment by JulienPeloton** [comment]
Thanks @cAbhi15 for the nice PR! Merging.

**Pull request merged**